pkg/requests/builder.go

```diff
@@ -0,0 +1,173 @@
+type Builder interface {
+	WithContext(context.Context) Builder
+	WithBody(io.Reader) Builder
+	WithMethod(string) Builder
+	WithHeader(http.Header) Builder
+	SetHeader(key, value string) Builder
+	Do() (*http.Response, error)
+	Into(interface{}) error
+	JSON() (*simplejson.Json, error)
+}
+
+type builder struct {
+	context  context.Context
+	method   string
+	endpoint string
+	body     io.Reader
+	header   http.Header
+	response *http.Response
+}
+
+// New provides a new Builder for the given endpoint.
+func New(endpoint string) Builder {
+	return &builder{
+		endpoint: endpoint,
+		method:   "GET",
+	}
+}
+
+// WithContext adds a context to the request.
+// If no context is provided, context.Background() is used instead.
+func (r *builder) WithContext(ctx context.Context) Builder {
+	r.context = ctx
+	return r
+}
+
+// WithBody adds a body to the request.
+func (r *builder) WithBody(body io.Reader) Builder {
+	r.body = body
+	return r
+}
+
+// WithMethod sets the request method. Defaults to "GET".
+func (r *builder) WithMethod(method string) Builder {
+	r.method = method
+	return r
+}
+
+// WithHeader replaces the request header map with the given header map.
+func (r *builder) WithHeader(header http.Header) Builder {
+	r.header = header
+	return r
+}
+
+// SetHeader sets a single header to the given value.
+// May be used to add multiple headers.
+func (r *builder) SetHeader(key, value string) Builder {
+	if r.header == nil {
+		r.header = make(http.Header)
+	}
+	r.header.Set(key, value)
+	return r
+}
+
+// Do performs the request and returns the response in its raw form.
+// If the request has already been performed, returns the previous result.
+// This will not allow you to repeat a request.
+func (r *builder) Do() (*http.Response, error) {
+	if r.response != nil {
+		// Request has already been done
+		return r.response, nil
+	}
+
+	// Must provide a non-nil context to NewRequestWithContext
+	if r.context == nil {
+		r.context = context.Background()
+	}
+
+	req, err := http.NewRequestWithContext(r.context, r.method, r.endpoint, r.body)
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %v", err)
+	}
+	req.Header = r.header
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("error performing request: %v", err)
+	}
+
+	r.response = resp
+	return resp, nil
+}
+
+// Into performs the request and attempts to unmarshal the response into the
+// the given interface. The response body is assumed to be JSON.
+// The response must have a 200 status otherwise an error will be returned.
+func (r *builder) Into(into interface{}) error {
+	resp, err := r.Do()
+	if err != nil {
+		return err
+	}
+
+	return Into(resp, into)
+}
+
+// JSON performs the request and attempts to unmarshal the response into a
+// simplejson.Json. The response body is assume to be JSON.
+// The response must have a 200 status otherwise an error will be returned.
+func (r *builder) JSON() (*simplejson.Json, error) {
```

Comment:
Related to to the `unmarshal` comment, should this be renamed to be clearer and in "verb" form to convey it is intended as a terminal method in the builder chain?

Otherwise cursory glance at the API might lead someone to thing `JSON()` sets up an encoding format or something and needs an action method chained after it.

Reply:
Yeah, this one should definitely be `UnmarshalJSON` at a minimum